src/cargo/ops/cargo_package/mod.rs

```diff
@@ -829,20 +829,31 @@ fn check_metadata(pkg: &Package, gctx: &GlobalContext) -> CargoResult<()> {
     );
 
     if !missing.is_empty() {
-        let mut things = missing[..missing.len() - 1].join(", ");
-        // `things` will be empty if and only if its length is 1 (i.e., the only case
-        // to have no `or`).
-        if !things.is_empty() {
-            things.push_str(" or ");
+        let should_warn = match pkg.publish() {
+            Some(registries) if registries.is_empty() => true, // publish = false, still warn
```

Comment:
If the warning is intended for crates.io and the package says it can't be published for crates.io, why warn in this case?